server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalMappedRareTerms.java

```diff
@@ -0,0 +1,132 @@
+                            List<B> buckets, long maxDocCount, BloomFilter bloom) {
+        // TODO is there a way to determine sum_other_doc_count and doc_count_error_upper_bound equivalents for rare based on bloom?
+        super(name, order, 0, 1, pipelineAggregators, metaData, format, 0, false, 0, buckets, 0);
+        this.maxDocCount = maxDocCount;
+        this.bloom = bloom;
+    }
+
+    public long getMaxDocCount() {
+        return maxDocCount;
+    }
+
+    BloomFilter getBloom() {
+        return bloom;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    InternalMappedRareTerms(StreamInput in, Bucket.Reader<B> bucketReader) throws IOException {
+        super(in, bucketReader);
+        maxDocCount = in.readLong();
+        bloom = new BloomFilter(in);
+    }
+
+    @Override
+    protected void writeTermTypeInfoTo(StreamOutput out) throws IOException {
+        super.writeTermTypeInfoTo(out);
+        out.writeLong(maxDocCount);
+        bloom.writeTo(out);
+    }
+
+    @Override
+    public InternalAggregation doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        Map<Object, List<B>> buckets = new HashMap<>();
+        InternalTerms<A, B> referenceTerms = null;
+        BloomFilter bloomFilter = null;
+
+        for (InternalAggregation aggregation : aggregations) {
+
+            // Unmapped rare terms don't have a bloom filter so we'll skip all this work
+            // and save some type casting headaches later.
+            if (aggregation.isMapped() == false) {
+                continue;
+            }
+
+            @SuppressWarnings("unchecked")
+            InternalTerms<A, B> terms = (InternalTerms<A, B>) aggregation;
+            if (referenceTerms == null && aggregation.getClass().equals(UnmappedRareTerms.class) == false) {
+                referenceTerms = terms;
+            }
+            if (referenceTerms != null &&
+                referenceTerms.getClass().equals(terms.getClass()) == false &&
+                terms.getClass().equals(UnmappedRareTerms.class) == false) {
+                // control gets into this loop when the same field name against which the query is executed
+                // is of different types in different indices.
+                throw new AggregationExecutionException("Merging/Reducing the aggregations failed when computing the aggregation ["
+                    + referenceTerms.getName() + "] because the field you gave in the aggregation query existed as two different "
+                    + "types in two different indices");
+            }
+            for (B bucket : terms.getBuckets()) {
+                List<B> bucketList = buckets.computeIfAbsent(bucket.getKey(), k -> new ArrayList<>());
+                bucketList.add(bucket);
+            }
+
+            if (bloomFilter == null) {
+                bloomFilter = ((InternalMappedRareTerms)aggregation).bloom;
```

Comment:
I think we should always merge into a new bloom filter. We have had issues in the past with reusing state from the shard responses so we tend to create new state for the reduce result now.

Reply:
++ will do